### PR TITLE
[WIP] 207 add the work of neidhart et al as a black box

### LINKED ISSUE #207: Add the work of Neidhart et al. as a black-box
https://academic.oup.com/genetics/article/198/2/699/5935970

### PR DESCRIPTION
**Pull request by RMichae1**
Add work by Aita et al. 
with additions of neidhart et al -> randomness from gen. Pareto instead of Gaussian , and "tunable" c.

Core method implemented as static method to an isolated black-box.

[WIP]
- [x] add tests logic
- [x] add tests batched function evaluations

### CODE REVIEW

**Comment by RMichae1** [comment]
Looks like the added tests are doing their job +1

**Comment by RMichae1** [comment]
Corrected black-box functionality and added tests for seeding and expected computed values.
Ready for review @miguelgondu .

Might have to rebase on current dev .

File: tox.ini [hunk 1]
```diff
@@ -98,7 +98,7 @@ commands=
     pytest {tty:--color=yes} -v -m "not slow and poli__dockstring" {posargs}
 
 [testenv:poli-rasp-py39]
-description = run the tests with pytest on the dockstring environment for poli
+description = run the tests with pytest on the RaSP environment for poli
```
Comment (miguelgondu) [inline tox.ini:101]: Nice catch, thanks!

File: src/poli/objective_repository/rmf_landscape/isolated_function.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    Raises
+    ------
+    AssertionError
+        If no wildtype sequence is provided.
+    """
+
+    def __init__(
+        self,
+        wildtype: List[str],
+        wt_val: Optional[float] = 0.0,
```
Comment (miguelgondu) [inline src/poli/objective_repository/rmf_landscape/isolated_function.py]: I think `Optional[Something]` is reserved as an abbreviation of `Something | None` (or `Union[Something, None]` in previous Python versions).

Comment (miguelgondu) [reply src/poli/objective_repository/rmf_landscape/isolated_function.py]: https://peps.python.org/pep-0484/#union-types

Comment (miguelgondu) [reply src/poli/objective_repository/rmf_landscape/isolated_function.py]: Also: feel free to include `from __future__ import annotations` and use the more contemporary annotations for unions.

File: src/poli/objective_repository/rmf_landscape/register.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    get_inner_function,
+    instance_function_as_isolated_process,
+)
+from poli.core.util.seeding import seed_python_numpy_and_torch
+from poli.objective_repository.rmf_landscape.information import rmf_info
+
+
+class RMFBlackBox(AbstractBlackBox):
+    """
+    TODO: docstring
```
Comment (miguelgondu) [inline src/poli/objective_repository/rmf_landscape/register.py]: Address this TODO.

File: src/poli/objective_repository/rmf_landscape/register.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    """
+
+    def __init__(
+        self,
+        wildtype: str,
+        wt_val: Optional[float] = 0.0,
```
Comment (miguelgondu) [inline src/poli/objective_repository/rmf_landscape/register.py]: Same comment as in `isolated_function.py`!

File: src/poli/objective_repository/rmf_landscape/register.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+        Creates RMF problem instance with specified parameters.
+    """
+
+    def get_setup_information(self) -> BlackBoxInformation:
+        return rmf_info
+
+    def create(
+        self,
+        wildtype: Union[List[str], str],
+        wt_val: Optional[float] = 0.0,
```
Comment (miguelgondu) [inline src/poli/objective_repository/rmf_landscape/register.py]: Same here!

File: src/poli/tests/registry/proteins/test_rmf.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+import pytest
+
+from poli import objective_factory
+from poli.objective_repository import AVAILABLE_PROBLEM_FACTORIES
+
+ref_aa_seq = "HPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMSTFKVLLCGAVLSRVDAGQEQLGRRIHYSQNDLVEYSPVTEKHLTDGMTVRELCSAAITMSDNTAANLLLTTIGGPKELTAFLHNMGDHVTRLDRWNPELNEAIPNDERDTTMPAAMATTLRKLLTGELLTLASRQQLIDWMEADKVAGPLLRSALPAGWFIADKSGAGERGSRGIIAALGPDGKPSRIVVIYTTGSQATMDERNRQIAEIGASLIKHW"
+
+
+@pytest.mark.poli__rmf
+def test_rmf_is_available():
```
Comment (miguelgondu) [inline src/poli/tests/registry/proteins/test_rmf.py]: This type of tests was unfortunately only relevant when we were doing isolation in the hacky way. Feel free to delete it. `"rmf_landscape"` will always be in `AVAILABLE_PROBLEM_FACTORIES`.

File: src/poli/tests/registry/proteins/test_rmf.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    y0_a = f_a(x0_a)
+    f_b, x0_b = problem_b.black_box, problem_b.x0
+    y0_b = f_b(x0_b)
+
+    y1_a = f_a(mutation_seq)
+    y1_b = f_b(mutation_seq)
+    # test equalities
+    assert all([x_a == x_b for x_a, x_b in zip(x0_a[0], x0_b[0])])
+    assert y0_a == y0_b
+    assert y1_a == y1_b  # value for mutated sequences equal
```
Comment (miguelgondu) [inline src/poli/tests/registry/proteins/test_rmf.py:84]: Quite thorough! I like it.

File: src/poli/tests/registry/proteins/test_rmf.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    y1 = f(mutation_seq)
+
+    rnd_state = np.random.default_rng(SEED)
+    ref_noise_0 = genpareto.rvs(f.kappa, size=1, random_state=rnd_state)
+    ref_noise_1 = genpareto.rvs(f.kappa, size=1, random_state=rnd_state)
+
+    # black-box value minus noisy component should be approximately mutational distance if c==1
+    assert np.isclose(np.round(y0 - ref_noise_0), 0)
+    assert np.isclose(np.round(y1 - ref_noise_1), -n_mutations)
+    f.terminate()
```
Comment (miguelgondu) [inline src/poli/tests/registry/proteins/test_rmf.py:118]: Also quite thorough. Great job!

**Review by miguelgondu — CHANGES_REQUESTED** [review]
Hey Richard! Great job, I left a couple of comments to address.

File: src/poli/objective_repository/rmf_landscape/isolated_function.py (hunk 2 above)
Comment (RMichae1) [reply src/poli/objective_repository/rmf_landscape/isolated_function.py]: Added future type annotations

File: src/poli/objective_repository/rmf_landscape/register.py (hunk 3 above)
Comment (RMichae1) [reply src/poli/objective_repository/rmf_landscape/register.py]: Added docstring.

File: src/poli/objective_repository/rmf_landscape/register.py (hunk 4 above)
Comment (RMichae1) [reply src/poli/objective_repository/rmf_landscape/register.py]: Addressed.

File: src/poli/objective_repository/rmf_landscape/register.py (hunk 5 above)
Comment (RMichae1) [reply src/poli/objective_repository/rmf_landscape/register.py]: Done.

File: src/poli/tests/registry/proteins/test_rmf.py (hunk 6 above)
Comment (RMichae1) [reply src/poli/tests/registry/proteins/test_rmf.py]: Removed this test.

**Review by miguelgondu — APPROVED** [review]
Looks great! Thanks for addressing the TO-DOs.

**Comment by RMichae1** [comment]
Thanks for the review!:)

**Pull request merged**